types/module.go

```diff
@@ -0,0 +1,209 @@
+package types
+
+import (
+	"encoding/json"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/spf13/cobra"
+	abci "github.com/tendermint/tendermint/abci/types"
+)
+
+// ModuleClient helps modules provide a standard interface for exporting client functionality
+type ModuleClient interface {
+	GetQueryCmd() *cobra.Command
+	GetTxCmd() *cobra.Command
+}
+
+//__________________________________________________________________________________________
+// AppModule is the standard form for basic non-dependant elements of an application module
+type AppModuleBasic interface {
```

Comment:
Is there some rationale behind these types? I.e. why isn't this defined on the `AppModule` interface? Is there a place where we use the `AppModuleBasic` but not `AppModule`?

Reply:
> Is there some rationale behind these types

short answer: _yes_ 
let me break it down for you: a lot of core module functionality is dependant on the module's keeper, many of the module keepers are dependant on each other, thus in order to access the full set of module functionality (aka the `AppModule` type) we need to define all the keepers/params-store/keys etc.   

However, there are several instances where simply module functionality is required but it would be totally unnecessary need to require to define all the keepers+junk (or actually impossible, because we're upstream in the code of where the keepers would even be defined)  - however we still want to allow for a high level pattern for modules to follow - for instance, such that we don't have to manually register all of the codecs for all the modules. 

so for this reason we _need_ to split out non-dependant (aka independant) functions

________
I'll beef up the code comment a bit to reflect this discussion

Reply:
Thats a great explanation. Thank you for taking the time there @rigelrozanski!

Reply:
k I added a decent explanation at the top of types/module.go going into this design choice a bunch more